stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
+     */
+    template<typename T1, typename T2>
+    inline void append_array_copy(const T1& x, T2& y) {
+      y = T2(x);
+    }
 
     /**
-     * Return the concatenation of two specified integer vectors in the
-     *   order of the arguments.
+     * This copies and casts the elements of x to y (by delegating to the Eigen
+     * template cast)
      *
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * @tparam Scalar type of source matrix
+     * @tparam Scalar Type of destination matrix
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x Source matrix
+     * @param y Target matrix
      */
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
+      y = x.template cast<T2>();
     }
 
     /**
-     * Return the concatenation of two vectors containing Matrices
-     *   of the same type in the order of the arguments.
+     * This function copies and casts the elements of std::vector x to std::vector y.
+     * First, y is resized to match the size of x, and then append_array_copy
+     * is recursively called on each element of x to copy it to an element of y.
      *
-     * @tparam T1 Type of element in Matrices in first vector
-     * @tparam T2 Type of element in Matrices in second vector
-     * @tparam R Row specification of matrices
-     * @tparam C Column specification of matrices
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
      */
-    template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template<typename T1, typename T2>
+    inline void append_array_copy(const std::vector<T1> &x,
+                                  std::vector<T2> &y) {
+      y.resize(x.size());
```

Comment:
What are the invariants here?  Is it possible to call this without `x` and `y` being the same size given your checks above?  Those just checked first elements, so are you anticipating ragged uses here or something (in which case those first element checks are wrong).